### PR TITLE
Add a new test to catch exit code failure

### LINKED ISSUE #29226: Add an integration test to verify spack exit code on the command line
In #25601 we removed a return statement from `main` accidentally, and we had no test that caught the issue. We need some test that can verify if failing commands are returning a non-zero exit code, since otherwise we'll get false positives in CI. 

The bug has been fixed in the meanwhile in #29223

### PR DESCRIPTION
**Pull request by alalazo**
fixes #29226

This introduces a new integration test that checks the return code of `spack unit-test` when it is supposed to fail.

This is to prevent bugs like the one introduced in #25601 in which CI didn't catch a missing return statement.

In retrospective it seems that the shell tests we have right now go through code paths which call `sys.exit` explicitly. This new test instead checks `spack unit-test` which relies on the return code from command invocation in case of errors.

### CODE REVIEW

File: lib/spack/spack/main.py [hunk 1]
```diff
@@ -868,7 +868,7 @@ def _main(argv=None):
         bootstrap_context = bootstrap.ensure_bootstrap_configuration()
 
     with bootstrap_context:
-        return finish_parse_and_run(parser, cmd_name, env_format_error)
+        finish_parse_and_run(parser, cmd_name, env_format_error)
```
Comment (alalazo) [inline lib/spack/spack/main.py]: I'm temporarily reintroducing the bug and checking if the shell tests now fail

Comment (alalazo) [reply lib/spack/spack/main.py]: Test failure is [here](https://github.com/spack/spack/runs/5359119769?check_suite_focus=true)

File: share/spack/qa/setup-env-test.sh [hunk 2]
```diff
@@ -200,3 +200,6 @@ despacktivate
 echo "Correct error exit codes for activate and deactivate"
 fails spack env activate nonexisiting_environment
 fails spack env deactivate
+
+echo "Correct error exit codes for unit-test when it fails"
+fails spack unit-test fail
```
Comment (scottwittenburg) [inline share/spack/qa/setup-env-test.sh:205]: @alalazo  This seems like a good way to test that command return statements translate to exit codes, nice!  I was unaware of this file at all, but now that I'm looking through it, it seems like the natural place for this test.

**Review by scottwittenburg — APPROVED** [review]
Nice, thanks!

**Review by becker33 — APPROVED** [review]
This is very clean
